### PR TITLE
[analyzer] Handle [[assume(cond)]] as __builtin_assume(cond)

### LINKED ISSUE #100762: [analyzer] Handle `[[assume(cond)]]` as `__builtin_assume(cond)`
We already handle `__builtin_assume` calls inside the `BuiltinFunctionChecker`.
We should also handle the attribute variant of this, e.g.:
https://compiler-explorer.com/z/oq6nY7eqE
```c++
void clang_analyzer_printState();
extern int arr[10];

void using_assume_attr(int x) {
    [[assume(x > 100)]]; // NullStmt with an attribute
    //clang_analyzer_printState();
    arr[x] = 404; // FIXME: we don't raise this!
}

void using_builtin(int x) {
    __builtin_assume(x > 100); // CallExpr
    //clang_analyzer_printState();
    arr[x] = 404; // caught!
}
```

### PR DESCRIPTION
**Pull request by vinay-deshmukh**
Resolves #100762 

Gist of the change:
1. All the symbol analysis, constraint manager and expression parsing logic was already present, but the previous code didn't "visit" the expressions within `assume()` by parsing those expressions, all of the code "just works" by evaluating the SVals, and hence leaning on the same logic that makes the code with `__builtin_assume` work
2. "Ignore" an expression from adding in CFG if it has side-effects ( similar to CGStmt.cpp (todo add link))
3. Add additional test case for ternary operator handling and modify CFG.cpp's VisitGuardedExpr code for `continue`-ing if the `ProgramPoint` is a `StmtPoint`

### CODE REVIEW

**Comment by vinay-deshmukh** [comment]
@steakhal 

Can you take a look at this PR when you have a chance?
Thanks!

**Comment by steakhal** [comment]
I simplified your implementation, and also identified a shortcoming of the `getSpecificAttr` that it didn't work with const attributes. I'll submit a fix for it separately to this PR.

The implementation looks pretty but there is a catch. And this is unfortunately a blocker issue.
Quoting https://eel.is/c++draft/dcl.attr.assume:

```c++
int divide_by_32(int x) {
  [[assume(x >= 0)]];
  return x/32;                  // The instructions produced for the division
                                // may omit handling of negative values.
}
int f(int y) {
  [[assume(++y == 43)]];        // y is not incremented
  return y;                     // statement may be replaced with return 42;
}
```

Because we eval the subexpressions of the assume arguments, their sideeffects sneaks in to the ExplodedGraph.
So after all, this approach might not work out that well.

We should have a look at what the codegen does with this attribute, or even with the builtin assume call - because we should do something very similar. We should gather the constraints and apply them, just like codegen does (I assume).
I think a suitable place to start reading would be CGStmt.cpp `CodeGenFunction::EmitAttributedStmt()`.

Basically, my idea now is more like: don't eval the expressions of the assume in the CFG, but only visit the AttributedStmt in the ExprEngine. When we would reach that, "do the same as codegen to get the constraints" and apply these constraints.

Alternatively, we could take a snapshot of the Store and Environment before we would evaluate any expressions of the assume attribute and "restore" those after we finished with the AttributedStmt. If we would do this, we would be careful of sinking/"merging" all the paths we ended up creating while evaluating this Stmt. This is a lot more complicated, so I'd definitely try the previous approach before even considering this.

File: clang/lib/StaticAnalyzer/Core/ExprEngineCXX.cpp [hunk 1]
```diff
@@ -1200,3 +1200,20 @@ void ExprEngine::VisitLambdaExpr(const LambdaExpr *LE, ExplodedNode *Pred,
...
+
+void ExprEngine::VisitAttributedStmt(const AttributedStmt *A,
+                                     ExplodedNode *Pred, ExplodedNodeSet &Dst) {
+  ExplodedNodeSet CheckerPreStmt;
+  getCheckerManager().runCheckersForPreStmt(CheckerPreStmt, Pred, A, *this);
+
+  ExplodedNodeSet EvalSet;
+  StmtNodeBuilder Bldr(CheckerPreStmt, EvalSet, *currBldrCtx);
+
+  if (const auto *AssumeAttr = getSpecificAttr<CXXAssumeAttr>(A->getAttrs())) {
```
Comment (vinay-deshmukh) [inline clang/lib/StaticAnalyzer/Core/ExprEngineCXX.cpp:1212]: @steakhal 
Q:
As this is a `if()` instead of `for(const auto *AssumeAttr: A->getAttrs())`, won't it ignore the `assume()` expressions _after_ the first one? Should we restore the for loop so the ExprEngine processes all `assume()` attributes?

i.e. this change: https://github.com/llvm/llvm-project/commit/7f1e341e59a52017d9f73ab09e1be45444536731

Before:
https://github.com/llvm/llvm-project/blob/daddb9e13db6ca8373dc7298d17aa36a03014aeb/clang/lib/Analysis/CFG.cpp#L469-L479

After:
https://github.com/llvm/llvm-project/blob/7f1e341e59a52017d9f73ab09e1be45444536731/clang/lib/StaticAnalyzer/Core/ExprEngineCXX.cpp#L1212-L1216

Comment (vinay-deshmukh) [reply clang/lib/StaticAnalyzer/Core/ExprEngineCXX.cpp:1212]: Same for the `for` -> `if` change in `CFG.cpp`

**Comment by steakhal** [comment]
I was thinking about the case, and I think it's okay to have state-splits in the subexpressions coming from an assume expression - given that the assume expression has no side effects.
This way we should have 2 paths after the assume expression (but still before the first if statement).
1) where `a > 10` is true
2) where `a > 10` is false, aka. `a <= 10`.

This is I think the current behavior of your proposed implementation. Could you demonstrate it in tests?

Something like this should prove it:
```c++
int ternary_in_assume(int a, int b) {
  [[assume(a > 10 ? b == 4 : b == 10)]];
  clang_analyzer_value(a); // we should have 2 dumps here.
  clang_analyzer_dump(b); // This should be either 4 or 10.
  if (a > 20) {
    clang_analyzer_dump(b + 100); // expecting 104
    return;
  }
  if (a > 10) {
    clang_analyzer_dump(b + 200); // expecting 204
    return;
  }
  clang_analyzer_dump(b + 300); // expecting 310
}
```

**Comment by vinay-deshmukh** [comment]
Aside: 

As I've been working on this issue, I've realized that it involves some complexities that might not make it an ideal "good first issue" (in my opinion). To keep labels consistent with the difficult level, if it is okay with you, it would be great if you could re-classify the original issue by removing the "good-first-issue" label as it might set incorrect expectations from new contributors.

File: clang/test/Analysis/cxx23-assume-attribute.cpp [hunk 2]
```diff
@@ -0,0 +1,49 @@
+// RUN: %clang_analyze_cc1 -std=c++11 -Wno-array-bounds -verify %s \
+// RUN:   -analyzer-checker=unix,core,alpha.security.ArrayBoundV2,debug.ExprInspection
+
+// TODO: fix the checker arguments
```
Comment (steakhal) [inline clang/test/Analysis/cxx23-assume-attribute.cpp:4]:
```suggestion
// RUN: %clang_analyze_cc1 -std=c++23 -analyzer-checker=core,debug.ExprInspection -verify %s
```

**Review by steakhal — COMMENTED** [review]
> As far as I know, to eliminate the warning emitted corresponding to the FIXME comment, we need to implement a new Checker for attributes; we can do this in a second PR to avoid scope creep for this issue/PR.

Agreed. I had a look at the exploded graph and look correct to me.

I think I'll come back to this PR, apply some really minor recommendations and I think we can land it.
Thank you so much for pushing for this so hard. I really appreciate it.

Expect me back on this in a couple of days.

**Comment by steakhal** [comment]
@vinay-deshmukh  WDYT of the current shape? Can we merge it?

**Review by steakhal — APPROVED** [review]
For me it looks good.

**Comment by vinay-deshmukh** [comment]
Looks good to me @steakhal ! Thanks for the review!

**Comment by steakhal** [comment]
I made the followup for this one in #120572

**Comment by vitalybuka** [comment]
> I made the followup for this one in #120572

Please severt and reland with followup. Followup is good way to go for trivial cases which can be landed with quick or no review

**Comment by thurstond** [comment]
I've reverted it in https://github.com/llvm/llvm-project/commit/2b9abf0db2d106c7208b4372e662ef5df869e6f1 to clear up some buildbots

**Comment by steakhal** [comment]
> I've reverted it in https://github.com/llvm/llvm-project/commit/2b9abf0db2d106c7208b4372e662ef5df869e6f1 to clear up some buildbots

Thanks! Ill have a look later.

**Comment by vinay-deshmukh** [comment]
Thanks @thurstond for the revert! Let me know if there's anything I can do here to help fix the issue!

**Comment by steakhal** [comment]
So we had the following assertion failure:
```
clang/lib/Analysis/ThreadSafety.cpp:529: LocalVariableMap::addDefinition: Assertion `!Ctx.contains(D)' failed.
```
On `libcxx/test/std/algorithms/alg.modifying.operations/alg.fill/pstl.fill_n.pass.cpp`.

I wish I could come back to this to see why that assert fails. I don't have the bandwidth. Maybe at some point in January.
